types/enzyme/index.d.ts

```diff
@@ -21,10 +21,22 @@ export class ElementClass extends Component<any, any> {
 /* These are purposefully stripped down versions of React.ComponentClass and React.StatelessComponent.
  * The optional static properties on them break overload ordering for wrapper methods if they're not
  * all specified in the implementation. TS chooses the EnzymePropSelector overload and loses the generics
+ *
+ * The Class1/2/3 is kind of ugly, but we need to allow any of the variations of no-constructor,
+ * one-arg constructor, and two-arg constructor. We can't do this with a single definition like
+ * `new(props?, context?)` because `props | undefined` is actually a wider type definition than
+ * just `constructor(props, context)`.
  */
-export interface ComponentClass<Props> {
-    new(props?: Props, context?: any): Component<Props, any>;
+export interface ComponentClass1<Props> {
+    new(): Component<Props, any>;
+}
+export interface ComponentClass2<Props> {
+    new(props: Props): Component<Props, any>;
+}
+export interface ComponentClass3<Props> {
+    new(props: Props, context: any): Component<Props, any>;
 }
+export type ComponentClass<Props> = ComponentClass1<Props> | ComponentClass2<Props> | ComponentClass3<Props>;
```

Comment:
How about naming interfaces like this?
```
ComponentClassConstructor<Props>
ComponentClassWithPropsConstructor<Props>
ComponentClassWithPropsAndContextConstructor<Props>
```

/cc @DovydasNavickas

Reply:
Agreed, I like that better. I'll push an update.

Reply:
Actually, I think a simpler fix is just removing the `?` from props. This is passing all of the existing enzyme-tests (plus the one I added that was breaking strict null checking):

```
export interface ComponentClass<Props> {
    new(props: Props, context?: any): Component<Props, any>;
}
```

My "guess we should handling matching a 0-arg constructor" was my ES6/TS newbie showing: if the component doesn't declare an explicit constructor, it gets an N-arg constructor based on its parent class. So we don't need to handle the 0-arg case.

And, AFAICT, components with both `constructor(props: ...)` and `constructor(props: ..., context: any)` work with the above `ComponentClass` with just `new(prop: Props, context?: any)`.

I'm actually not sure why the `context?` can be optional, but `props?` cannot be.